src/darwin/Framework/CHIP/MTRDeviceController.mm

```diff
@@ -654,6 +660,54 @@ - (void)setPairingDelegate:(id<MTRDevicePairingDelegate>)delegate queue:(dispatc
     });
 }
 
+- (void)setNocChainIssuer:(id<NOCChainIssuer>)nocChainIssuer
+{
+    dispatch_sync(_chipWorkQueue, ^{
+        self->_operationalCredentialsDelegate->SetNocChainIssuer(nocChainIssuer);
+    });
+}
+
+- (uint32_t)onNOCChainGeneration:(NSData *)operationalCertificate
+         intermediateCertificate:(NSData *)intermediateCertificate
+                 rootCertificate:(NSData *)rootCertificate
+                             ipk:(NSData *)ipk
+                    adminSubject:(uint64_t)adminSubject
+{
+    // use ipk and adminSubject from CommissioningParameters if not passed in
+    chip::Controller::CommissioningParameters commissioningParams = _autoCommissioner->GetCommissioningParameters();
+
+    chip::Optional<chip::Crypto::AesCcm128KeySpan> ipkOptional;
+    uint8_t ipkValue[chip::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES];
+    chip::Crypto::AesCcm128KeySpan ipkTempSpan(ipkValue);
+    chip::ByteSpan ipkBytes = AsByteSpan(ipk);
+    if (ipk != nil) {
+        VerifyOrReturnValue(ipkBytes.size() == sizeof(ipkValue), CHIP_ERROR_INTERNAL.AsInteger());
+        memcpy(&ipkValue[0], ipkBytes.data(), ipkBytes.size());
+        ipkOptional.SetValue(ipkTempSpan);
+    } else if (commissioningParams.GetIpk().HasValue()) {
+        ipkOptional.SetValue(commissioningParams.GetIpk().Value());
+    }
+
+    chip::Optional<chip::NodeId> adminSubjectOptional;
+    if (adminSubject == chip::kUndefinedNodeId) {
+        // if no value passed in, use value from CommissioningParameters
+        adminSubject = commissioningParams.GetAdminSubject().ValueOr(chip::kUndefinedNodeId);
+    } else {
+        adminSubjectOptional.SetValue(adminSubject);
+    }
+
+    __block CHIP_ERROR err = CHIP_NO_ERROR;
+    dispatch_sync(_chipWorkQueue, ^{
```

Comment:
Why is this a sync dispatch?  This is a fundamentally async operation, no?

Reply:
The call to NOCChainGenerated ends up calling DeviceCommissioner::OnDeviceNOCChainGeneration which ultimately send the AddNoc message to the target device. The call will then return, no blocking is involved. Is there a strong reason to make this async still?

Reply:
Added documentation to state this in the code.